### ISSUE TITLE
Dementia It's A State Of Mind

### ISSUE THREAD

**Issue by IllustriousProsperus** [narzoul/DDrawCompat#402]
This game is more unstable than dynamite, without ddraw the game requires 256 colors and 640 x 480 resolution. With ddrawcompat
as soon as you start the game there is only a black screen but you can see the cursor. Pressing "alt f4" the message to exit the game appears and if you press "alt tab" you can see the menu. Click on " new game" the game freezes but it is always possible to exit the game with the exit confirmation by pressing "alt f4".

cursor in the middle

![black](https://github.com/user-attachments/assets/d58e680a-252b-42e6-b2db-ea75f1ee0d1a)

main menu

![menu](https://github.com/user-attachments/assets/b0d77191-2387-4698-97a9-3d0f492bf945)

crash

![tnt](https://github.com/user-attachments/assets/4c87a703-a90a-4f60-b19c-67fb149b0b5b)

log only, crash dump not available

https://mega.nz/folder/rA9FiI7I#a2ZwJf-lPrbGY6ePUo7YHQ

**Comment by BEENNath58** [comment]
I tested the game on DxWnd today. The crash here was solved by the Exception Handler though I didn't bother seeing what the crash was.

With DDrawCompat, I can emulate the first 2 photos. The crash photo is probably not needed since the palette can be randomly corrupted which cause a crash

**Comment by narzoul** [comment]
Ah, the palette issue was simpler than I expected. Here's a patch for that:
[ddraw.zip](https://github.com/user-attachments/files/19720726/ddraw.zip) ([diff.txt](https://github.com/user-attachments/files/19720727/diff.txt) compared to v0.6.0)

I did not get a crash yet. Is that still an issue?

**Comment by BEENNath58** [comment]
This one works and also doesn't crash.

There is an issue, for example if I Alt+F4 I get the exit prompt. If I press NO, I expect the game to return but it is just a black screen

**Comment by narzoul** [comment]
> There is an issue, for example if I Alt+F4 I get the exit prompt. If I press NO, I expect the game to return but it is just a black screen

Works fine for me.

**Comment by narzoul** [comment]
> Well it started crashing again.

Please send debug logs and a crash dump. I haven't had any crashes yet.

**Comment by IllustriousProsperus** [comment]
> > Well it started crashing again.
> 
> Please send debug logs and a crash dump. I haven't had any crashes yet.

The save screen has the wrong colors but I managed to load the game. I used a save taken from the 86box vm, I think it's the "animations" that crash the game, in fact it crashed trying to open the "braventory". The game needs the argument 

"CRUX.EXE -h"

i wonder if there are others...

In the video the game crashes but sometimes the classic "unexpected error" error appears (but more rarely than usual...)

https://github.com/user-attachments/assets/5ba2bc5a-8595-4e3d-bef1-52943f71cc5e

![Image](https://github.com/user-attachments/assets/9079b236-7af9-4e58-941c-c58a25bb9ebe)

https://mega.nz/file/DBkkjTCD#Xktm1DC0x67KeIxCEEiu-vSV9AorkNwf99Hrc7f5gL0

**Comment by narzoul** [comment]
> I managed to capture a clip where after Alt+F4 the scene doesn't return

I had this happen once a few days ago, but now that I'm trying to reproduce it to debug the issue, it refuses to fail even after dozens of alt-tab attempts of course. Anyway, from what I remember it wasn't just a rendering issue, the whole game was frozen. But until I can reproduce it, I can't really check if it has something to do with DDrawCompat.

> The save screen has the wrong colors

It's a side effect of proper GDI interop, which makes the save dialogue also run in 256 color mode, and those modern icons are probably not prepared to be displayed like that.

I might revert GDI interworking for all top-level windows or something in the future, or at least for dialog boxes. I'll think about it...

> In the video the game crashes but sometimes the classic "unexpected error" error appears (but more rarely than usual...)

An "Errore inatteso" message box appeared in your logs too. My guess is it doesn't always appear on screen because it doesn't have a parent window, so it might get covered by the game's main window or something like that. It was followed by another MessageBox with seemingly invalid/useless parameters (no caption or text): `MessageBoxA(null, null, null, 65536)`

The crash followed this, but I suspect it's just due to bad error handling at this point, and the problem happened somewhere earlier before the "Errore inatteso" message, but there is seemingly nothing wrong from the logs prior to that. So I think it's unlikely to have anything to do with DDrawCompat. But just like the alt-tab issue, I cannot reproduce it on my end now, so I can't check it with a debugger.

**Comment by IllustriousProsperus** [comment]
Here is another crash dump, in any case your ddraw is the only one that is able to "start" the game 100% of the time, on 86box with win95fe there are no problems, with win95se you already notice several crashes and with win98fe the game does not work at all.

https://mega.nz/file/zAtjEKZA#AspLR3BXYRYi0xPXWgVW-4b2mA28RG39ael83Lq2RN8

**Comment by narzoul** [comment]
Same as last time, the game errors out, but not from any DirectDraw issue apparently. I don't think I can help with this. Perhaps the Italian version is different, as I still cannot even reproduce it with the English version.

The game seems to write some logs into ADVENT.OUT. You could try to check if it writes some more specific error message there, but I doubt it.

**Comment by BEENNath58** [comment]
Umm I am not sure how to get the best DUMP. Since the dumping ends after the application crashes.
Now the game works yet again!

**Comment by narzoul** [comment]
> Umm I am not sure how to get the best DUMP.

Same way as I wrote about in your other issue: https://github.com/narzoul/DDrawCompat/issues/459#issuecomment-2802590081

If you can reproduce it again, make sure you attach also the debug logs, not just the crash dump.

**Comment by narzoul** [comment]
Unfortunately those aren't debug logs. But based on where it crashed, I don't think I'll see anything more useful from debug logs anyway. It's addressing some internal array at the crash site, so it's unlikely to be related to a DirectDraw surface lock or anything else related to DirectDraw.

**Comment by BEENNath58** [comment]
I used minidump, does this say the same thing now:

[program.zip](https://github.com/user-attachments/files/20012247/program.zip)

**Comment by narzoul** [comment]
It crashed at the same place. But the problem was not with the crash dump, but with the logs. I.e. you haven't enabled LogLevel=debug, so I don't see what was happening before the crash.

**Comment by narzoul** [comment]
Yeah, it's fine now, But it's the same issue that IllustriousProsperus has, the game throws some "Unexpected error" message box, then some other (empty/invalid) message box and crashes afterwards. But the crash is not very interesting because the game has already got into a bad state long before, I wouldn't bother trying to investigate that.

But there is also nothing wrong with the rendering in your logs prior to the error, so I don't think the issue is related to DirectDraw.

> I can't find ADVENT.OUT

I think it's because you installed the game in Program Files. I also don't get the file if I put the game in Program Files (it doesn't even appear in VirtualStore it seems). Try moving it to some non-protected directory, then see if there is anything useful in it when the game throws an error again.

**Comment by narzoul** [comment]
> I think it's because you installed the game in Program Files. I also don't get the file if I put the game in Program Files (it doesn't even appear in VirtualStore it seems). Try moving it to some non-protected directory, then see if there is anything useful in it when the game throws an error again.

Never mind, the file is still created for me, but in the original install directory instead of where I copied the game. I'm not sure how the game determines that.

There is also a data\ERROR.LOG file that may be useful. There is only one exception logged there for me, from a couple weeks ago, I think it might be the one time when I had a similar error message, although for me it was from clicking some button in the main menu.

```
An unhandled exception has occured.
Call Stack:
ani_to_mem(char *name,int type,int ani_slot,int read_delay)
ani_add_by_name(char *name,int read_delay)
ani_add_by_num(unsigned num,int type,int read_delay)
run_prog(unsigned num,int id)
run_scn(char *scr_name)
int execution_thread(int val)
```

**Comment by BEENNath58** [comment]
I assume you mean
C:\Users\%username%\AppData\Local\VirtualStore 

In that case it doesn't contain any ADVENT.OUT. Just some files from my old Midtown Madness testing.

My game was installed in E:\Program Files\, not a system drive, so it also makes little sense for it to go to VirtualStore

ERROR.LOG is same as you.

Other than figuring out why ADVENT.OUT wasn't created, I think there is nothing else to do here

**Comment by narzoul** [comment]
The palette issue was fixed in v0.7.0. The crash is out of scope.
